tool/rbinstall.rb

```diff
@@ -705,28 +705,38 @@ def remove_prefix(prefix, string)
     end
   end
 
-  class UnpackedInstaller < Gem::Installer
-    module DirPackage
-      def extract_files(destination_dir, pattern = "*")
-        path = File.dirname(@gem.path)
-        return if path == destination_dir
-        File.chmod(0700, destination_dir)
-        mode = pattern == "bin/*" ? $script_mode : $data_mode
-        spec.files.each do |f|
-          src = File.join(path, f)
-          dest = File.join(without_destdir(destination_dir), f)
-          makedirs(dest[/.*(?=\/)/m])
-          install src, dest, :mode => mode
-        end
-        File.chmod($dir_mode, destination_dir)
-      end
+  class DirPackage
+    attr_accessor :spec
+
+    attr_accessor :dir_mode
+    attr_accessor :prog_mode
+    attr_accessor :data_mode
+
+    def initialize(spec)
+      @spec = spec
+      @src_dir = File.dirname(@spec.loaded_from)
     end
 
-    def initialize(spec, *options)
-      super(spec.loaded_from, *options)
-      @package.extend(DirPackage).spec = spec
+    def extract_files(destination_dir, pattern = "*")
+      path = @src_dir
+      return if path == destination_dir
+      File.chmod(0700, destination_dir)
+      mode = pattern == File.join(@spec.bindir, '*') ? prog_mode : data_mode
+      spec.files.each do |f|
```

Comment:
Since you are already moving the ``extract_files`` method:
- Is there a reason to have a ``path`` local variable instead of using ``@src_dir``?
- You could use ``@spec`` at line 725, ``spec.files.each do |f|``.

Reply:
Initially, I just moved the method, keeping it mostly unmodified to keep it easy to follow the changes during a review. So there are definitely things to improve. E.g. repetitive call to `without_destdir(destination_dir)` except the issues you mentioned. I can take a look into these.

Reply:
> Could spec be replaced by `@spec` at line 725, `spec.files.each do |f|`?

Since the `attr_reader :spec` is required (actually it can be just reader, I'm going to change it), I think I should rather change the line 724 to use the `spec` method instead of the instance variable.

Reply:
Now the `attr_reader :spec` is used everywhere and I have dropped the `path` variable in favor of `@src_dir`.

I am still unsure about the `without_destdir(destination_dir)` becase 1) the naming is hard and I can't figure out a suitable variable name 2) I still consider changing `:install_dir => with_destdir(gem_dir),` to just `:install_dir => gem_dir,` and fixing the rest.

Reply:
At the end, I have refactored the `without_destdir(destination_dir)` to be cached as `destdir`. All your concerns should be resolved now.